workflow/artifacts/s3/s3.go

```diff
@@ -180,7 +180,27 @@ func (s3Driver *ArtifactDriver) Delete(artifact *wfv1.Artifact) error {
 		if err != nil {
 			return err
 		}
-		return s3cli.Delete(artifact.S3.Bucket, artifact.S3.Key)
+
+		isDir, err := s3cli.IsDirectory(artifact.S3.Bucket, artifact.S3.Key)
+		if err != nil {
+			return fmt.Errorf("failed to test if %s is a directory: %v", artifact.S3.Key, err)
+		}
+
+		if !isDir {
+			return s3cli.Delete(artifact.S3.Bucket, artifact.S3.Key)
+		} else {
+			keys, err := s3cli.ListDirectory(artifact.S3.Bucket, artifact.S3.Key)
+			if err != nil {
+				return fmt.Errorf("unable to list files in %s: %s", artifact.S3.Key, err)
+			}
+			for _, objKey := range keys {
+				err = s3cli.Delete(artifact.S3.Bucket, objKey)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		return nil
```

Comment:
```suggestion
		}
		
		keys, err := s3cli.ListDirectory(artifact.S3.Bucket, artifact.S3.Key)
		if err != nil {
			return fmt.Errorf("unable to list files in %s: %s", artifact.S3.Key, err)
		}
		for _, objKey := range keys {
			err = s3cli.Delete(artifact.S3.Bucket, objKey)
			if err != nil {
				return err
			}
		}
	}
	return nil
```

the else is unnecessary with the early return, so let's reduce nesting to reduce code complexity

see also other places where I've refactored code to reduce nesting: #11596, #12146, etc